Add BYPASS CACHE and USING TIMEOUT capabilities to mapper

Bouncheck opened the pull request:
Now Select, Insert and Update annotations used in DAOs can specify (usingTimeout = "durationString") in order to use Scylla CQL extension timeouts.

Select queries can also use (bypassCache = true).

Bouncheck committed:
Add BYPASS CACHE and partial USING TIMEOUT to mapper

Now Select, Insert and Update annotations used in DAOs
can specify (usingTimeout = "durationString") in order to use
Scylla CQL extension timeouts.

Select annotations can specify (bypassCache = true).

Fixes #168

Bouncheck commented:
My sincerest apologies to owners of Select, Insert and Update accounts that I've accidentally tagged. Feel free to review though.

Bouncheck commented:
I've decided to go with "usingTimeout" to make it more distinct from `@StatementAttributes(timeout)` which AFAIK is a driver-side timeout.

mykaul commented:
Ping - any updates to this PR?

roydahan commented:
@Bouncheck / @avelanarius what is this one waiting for?
Can we get a peer review in parallel?

roydahan commented:
Ping.

Lorak-mmk commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java:
```diff
@@ -103,6 +104,45 @@ protected void maybeAddTimestamp(String timestamp, MethodSpec.Builder methodBuil
+        methodBuilder.addCode(".$L($T.$L($S))", "usingTimeout", CqlDuration.class, "from", timeout);
+      }
+    }
+  }
```
Please use early return in case `timeout.isEmpty()` it would simplify the code - right now whole function has additional level of indentation that is unnecessary

Bouncheck replied:
Addressed in new version

Lorak-mmk commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java:
```diff
@@ -103,6 +104,45 @@ protected void maybeAddTimestamp(String timestamp, MethodSpec.Builder methodBuil
+                  "Invalid "
+                      + "USING TIMEOUT"
+                      + " value: "
+                      + "'%s' is not a valid placeholder, the generated query will probably fail",
```
Why not throw an exception in this case?
Failing early and rejecting such identifiers wold decrease probability of user bugs.

Bouncheck replied:
I followed the template of the rest of the code.
I don't see the reason why not to throw here so I've changed it to error.

Lorak-mmk commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/dao/DaoMethodGenerator.java:
```diff
@@ -103,6 +104,45 @@ protected void maybeAddTimestamp(String timestamp, MethodSpec.Builder methodBuil
+                      + " value: "
+                      + "'%s' is not a bind marker name and can't be parsed as a CqlDuration "
+                      + "either, the generated query will probably fail",
+                  timeout);
```
Again, why not throw in this case?

Lorak-mmk replied:
Ok I see that this method is pretty much copied from `maybeAddSimpleClause` which does the same thing.
I wonder if there is any reason for it, do you know? If there isn't maybe we should do better and throw?

Another question: why not use `maybeAddSimpleClause`?

Bouncheck replied:
`maybeAddSimpleClause` does almost what we want here but it expects arguments to be subclasses of Number. I've duplicated it and adjusted to parse CqlDurations for the timeouts. It also has a slightly different format. I've changed the `.warn` to `.error`.

Lorak-mmk commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/UsingTimeoutIT.java:
```diff
@@ -0,0 +1,117 @@
+import org.junit.rules.TestRule;
+
+@Category(ParallelizableTests.class)
+@CassandraSkip
```
Why skip Cassandra? Doesn't it support `USING TIMEOUT`?

Bouncheck replied:
AFAIK it's a [Scylla's CQL extension](https://enterprise.docs.scylladb.com/stable/cql/cql-extensions.html#using-timeout)

Lorak-mmk commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/UsingTimeoutIT.java:
```diff
@@ -0,0 +1,117 @@
+
+@Category(ParallelizableTests.class)
+@CassandraSkip
+@ScyllaRequirement(minOSS = "3.1.0", minEnterprise = "2020.1.0")
```
Did you verify those bounds or is it just a guess?

Bouncheck replied:
I'm pretty sure I've verified by checking labels of specific commits in Scylla.

Bouncheck replied:
I had mixed up versions here but it's correct now. I've added description for context.

Lorak-mmk reviewed (changes requested):
`USING TIMEOUT` is added to select insert and update - what about delete and batch?

PR title mentions `BYPASS CACHE` and I see those changes in the code, but there is no mention in PR description

Tests:
- I don't see tests for `BYPASS CACHE`. I know it's hard to test it's behavior (because it only affects performance) but you could at least test if generated query is correct and executed by Scylla
- IIUC tests only verify that negative timeouts result in specific exception. I don't know much about mapper in Java Driver, but I'd expect tests that 
     - verify produced query strings
     - execute correct statements with timeout
     - set timeout to very low value to verify it actually works

Bouncheck commented:
Thanks for the review, I see good points being raised. It has been a while so i may need to retrace some decisions I've made, but I'll try to address everything.

Bouncheck commented:
> USING TIMEOUT is added to select insert and update - what about delete and batch?

I've limited this change to queries mentioned in #168. Delete does not seem to support using timeout in QueryBuilder yet. I haven't seen batch and truncate used anywhere in mapper.

> PR title mentions BYPASS CACHE [...], but there is no mention in PR description

Added mention in description.

> Tests

Added tests that verify generated code,
Added integration tests that run correct statements with valid timeouts,
Added tests with invalid timeouts and valid but too low timeout,
Duplicated some integration tests for Select annotation but with added bypass cache so that at least we know it does not break
